OCPBUGS-51261:controller: track rte configmap state

Tal-or opened the pull request:
Kubelet controller is responsible to track `KubeletConfig` objects and
generate a `ConfigMap` data from them that will be served to RTE.

NROP controller is responsible to track the `ConfigMap` creation,
feed it to RTE and restart RTE in case the `ConfigMap` data has changed.

The latter is broken, and this PR is fixing it.

In addition the `ConfigMap` should be per `DaemonSet`/`NodeGroup`, while
previously it was common at the CR level which is wrong.

Because cluster might have multiple node groups which are configured
slightly different.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
controller: track rte configmap state

Kubelet controller is responsible to track `KubeletConfig` object and
generate a `ConfigMap` data from it that will be served for RTE.

NROP controller is responsible to track the `ConfigMap` creation,
feed it to RTE and restart RTE in case the `ConfigMap` data has changed.

The later is broken, and this commit is fixing it.

In addition the `ConfigMap` should be per `DaemonSet`/`NodeGroup`, while
previously it was common at the CR level which is wrong.

Because cluster might have multiple node groups which are configured
slightly different.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
unit-test: validate configmap hash

The test validates that the configmap hash is being added to the
daemonset annotation, and that it differ between different daemonsets.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
configmap: cleanup

remove unused code.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani commented:
conceptually LGTM. Please fix CI errors

ffromani commented on pkg/objectstate/rte/rte.go:
```diff
@@ -407,3 +416,20 @@ func getObject(ctx context.Context, cli client.Client, key client.ObjectKey, obj
+func findConfigMapState(name, namespace string, configMaps map[string]configMapManifest) *corev1.ConfigMap {
+	var desiredConfigMap *corev1.ConfigMap
+	existingConfigMap, ok := configMaps[name]
+	errFormat := fmt.Sprintf("failed to find config map %s/%s", namespace, name)
```
it seems we can log this in few flows, but all the flows should be mutually exclusive. So I'd avoid this and inline the formatting in the logging sites.

ffromani commented on internal/controller/numaresourcesoperator_controller.go:
```diff
@@ -753,6 +744,11 @@ func daemonsetUpdater(poolName string, gdm *rtestate.GeneratedDesiredManifest) e
 			klog.V(5).InfoS("DaemonSet update: selinux options", "container", manifests.ContainerNameRTE, "context", selinux.RTEContextType)
 		}
 	}
+	if gdm.ConfigMap != nil {
```
on which cases we do expect `gdm.ConfigMap` to be nil?

ffromani replied:
ok, makes sense

ffromani commented on pkg/objectstate/rte/nodegroup.go:
```diff
@@ -80,11 +89,27 @@ func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []obje
+	case existingConfigMap.configMapError != nil:
+		klog.ErrorS(existingConfigMap.configMapError, errFormat.Error())
+	default:
+		desiredConfigMap = existingConfigMap.configMap
```
do we need a DeepCopy here?

Tal-or replied:
No. We're not mutating the config map.
All the controller is doing is exercise a hash function on the data

ffromani commented on pkg/objectstate/rte/machineconfigpool.go:
```diff
@@ -96,11 +105,27 @@ func (obj machineConfigPoolFinder) FindState(mf Manifests, tree nodegroupv1.Tree
 
+		var desiredConfigMap *corev1.ConfigMap
+		existingConfigMap, ok := obj.em.configMaps[generatedName]
+		errFormat := fmt.Errorf("failed to find config map %s/%s", mf.Core.DaemonSet.Namespace, generatedName)
```
please inline in the calling site

ffromani commented on pkg/objectstate/rte/nodegroup.go:
```diff
@@ -80,11 +89,27 @@ func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []obje
 
+	var desiredConfigMap *corev1.ConfigMap
+	existingConfigMap, ok := obj.em.configMaps[generatedName]
+	errFormat := fmt.Errorf("failed to find config map %s/%s", mf.Core.DaemonSet.Namespace, generatedName)
```
please inline in the calling site

ffromani commented:
this work should probably be backported down to 4.16

Tal-or commented:
/cherry-pick release-4.18

Tal-or commented:
> so `ConfigMap` is already an exception in the context of this controller. We don't own it, we don't reconcile it, we "just" need to read it. So, idea: why don't we read it _and_ store its hash in the internal data structures vs storing the object and computing when needed? this
> 
> 1. will further reinforce the exceptional treatment
> 2. will remove any ambiguity about ownerhip, mutation or not

Sure this is an option. I used the above approach simply because it was aligned with the existing architecture

ffromani commented:
> > so `ConfigMap` is already an exception in the context of this controller. We don't own it, we don't reconcile it, we "just" need to read it. So, idea: why don't we read it _and_ store its hash in the internal data structures vs storing the object and computing when needed? this
> > 
> > 1. will further reinforce the exceptional treatment
> > 2. will remove any ambiguity about ownerhip, mutation or not
> 
> Sure this is an option. I used the above approach simply because it was aligned with the existing architecture

thanks, looks better IMO and almost ready to go (gonna review shortly).
My general rule is: code which behaves the same (or kinda the same) should look the same (or kinda the same). Code which behaves different can and should look different. So in this case looking different is a feature IMO.

ffromani commented on pkg/objectstate/rte/nodegroup.go:
```diff
@@ -74,6 +82,8 @@ func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []obje
 	desiredDaemonSet := mf.Core.DaemonSet.DeepCopy()
 	desiredDaemonSet.Name = generatedName
+	// it's possible that the hash will be empty if kubelet controller hasn't created a configmap
+	rteupdate.DaemonSetHashAnnotation(desiredDaemonSet, existingDaemonSet.rteComputedConfigHash)
```
I'm thinking aloud. Should we set it to "" (empty string) or should we set it only if `!= ""`? What are the pros and cons in both scenarios?

ffromani replied:
do we log when we change annotation? we probably should (V>=4)

Tal-or replied:
if kubeletconfig has been completely deleted from the cluster, we should be able to signal that correctly.
setting the annotation value to `""` gives us what we need.
It will cause RTE pods to restart and they'll reflect the right behavior as in the case of no kubeletconfig configuration at all.

ffromani replied:
good point, it's worthy to keep it. Please do *NOT* resubmit just to add logs

ffromani commented on pkg/objectstate/rte/machineconfigpool.go:
```diff
@@ -21,17 +21,20 @@ import (
 	"github.com/openshift-kni/numaresources-operator/pkg/objectstate"
 	"github.com/openshift-kni/numaresources-operator/pkg/objectstate/compare"
 	"github.com/openshift-kni/numaresources-operator/pkg/objectstate/merge"
+	rteupdate "github.com/openshift-kni/numaresources-operator/pkg/objectupdate/rte"
```
ok, this is the only remaining item (extra logging is a nice bonus but not required).
Conceptually, `objectstate` sits below in the stack with respect to `objectupdate`:

```
controller
objectupdate
objectstate
client
```
IOW, objectstate can't import objectupdate, but objectupdate can call objectstate (controller can and do import them directly).
We can bypass layers from up to down, but we should never break layering from the bottom to the up

Tal-or replied:
And that's why I preferred passing the configmap, because it is more align with current flow. Anyway I'll fix that, I'll find a way to update the ds later in the code

ffromani replied:
I may be wrong but  I think there should be a reasonnably easy way to not break the layering and yet only carry to hash around

ffromani reviewed (commented):
ok, last comment inline, but is worthy. If you add the logs (V>=4) we were talking about it's a nice bonus.

ffromani reviewed (commented):
the main problem here is we push updates in the state extraction code which was a bad idea in the first place back in time `daemonsetUpdater`. It should have been cleaner since the beginning.

Now's probably late to clean the board - we should wait 4.20 for it, if ever, so let's try to have a more targeted fix.

ffromani reviewed (commented):
something like this should fit more the existing (very debatable, see #1217) layout:
```
diff --git a/internal/controller/numaresourcesoperator_controller.go b/internal/controller/numaresourcesoperator_controller.go
index a10433bc..c44e8c21 100644
--- a/internal/controller/numaresourcesoperator_controller.go
+++ b/internal/controller/numaresourcesoperator_controller.go
@@ -758,6 +758,9 @@ func daemonsetUpdater(poolName string, gdm *rtestate.GeneratedDesiredManifest) e
 			klog.V(5).InfoS("DaemonSet update: selinux options", "container", manifests.ContainerNameRTE, "context", selinux.RTEContextType)
 		}
 	}
+
+	// set hash here
+
 	return nil
 }
 
diff --git a/pkg/objectstate/rte/machineconfigpool.go b/pkg/objectstate/rte/machineconfigpool.go
index 618319db..fa0a9140 100644
--- a/pkg/objectstate/rte/machineconfigpool.go
+++ b/pkg/objectstate/rte/machineconfigpool.go
@@ -75,12 +75,14 @@ func (obj machineConfigPoolFinder) FindState(mf Manifests, tree nodegroupv1.Tree
 	var ret []objectstate.ObjectState
 	for _, mcp := range tree.MachineConfigPools {
 		var existingDs client.Object
+		var rteConfigHash string
 		var loadError error
 
 		generatedName := objectnames.GetComponentName(obj.instance.Name, mcp.Name)
 		existingDaemonSet, ok := obj.em.daemonSets[generatedName]
 		if ok {
 			existingDs = existingDaemonSet.daemonSet
+			rteConfigHash = existingDaemonSet.rteComputedConfigHash
 			loadError = existingDaemonSet.daemonSetError
 		} else {
 			loadError = fmt.Errorf("failed to find daemon set %s/%s", mf.Core.DaemonSet.Namespace, mf.Core.DaemonSet.Name)
@@ -101,6 +103,7 @@ func (obj machineConfigPoolFinder) FindState(mf Manifests, tree nodegroupv1.Tree
 			MachineConfigPool:     mcp.DeepCopy(),
 			NodeGroup:             tree.NodeGroup.DeepCopy(),
 			DaemonSet:             desiredDaemonSet,
+			RTEComputedConfigHash: rteConfigHash,
 			IsCustomPolicyEnabled: annotations.IsCustomPolicyEnabled(tree.NodeGroup.Annotations),
 		}
 
diff --git a/pkg/objectstate/rte/nodegroup.go b/pkg/objectstate/rte/nodegroup.go
index 902c8e6d..c4adfff6 100644
--- a/pkg/objectstate/rte/nodegroup.go
+++ b/pkg/objectstate/rte/nodegroup.go
@@ -59,6 +59,7 @@ func (obj nodeGroupFinder) UpdateFromClient(ctx context.Context, cli client.Clie
 func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []objectstate.ObjectState {
 	var ret []objectstate.ObjectState
 	var existingDs client.Object
+	var rteConfigHash string
 	var loadError error
 
 	poolName := *tree.NodeGroup.PoolName
@@ -67,6 +68,7 @@ func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []obje
 	existingDaemonSet, ok := obj.em.daemonSets[generatedName]
 	if ok {
 		existingDs = existingDaemonSet.daemonSet
+		rteConfigHash = existingDaemonSet.rteComputedConfigHash
 		loadError = existingDaemonSet.daemonSetError
 	} else {
 		loadError = fmt.Errorf("failed to find daemon set %s/%s", mf.Core.DaemonSet.Namespace, mf.Core.DaemonSet.Name)
@@ -85,6 +87,7 @@ func (obj nodeGroupFinder) FindState(mf Manifests, tree nodegroupv1.Tree) []obje
 		MachineConfigPool:     nil,
 		NodeGroup:             tree.NodeGroup.DeepCopy(),
 		DaemonSet:             desiredDaemonSet,
+		RTEComputedConfigHash: rteConfigHash,
 		IsCustomPolicyEnabled: annotations.IsCustomPolicyEnabled(tree.NodeGroup.Annotations),
 	}
 
diff --git a/pkg/objectstate/rte/rte.go b/pkg/objectstate/rte/rte.go
index c13a4f5f..43df352e 100644
--- a/pkg/objectstate/rte/rte.go
+++ b/pkg/objectstate/rte/rte.go
@@ -54,8 +54,9 @@ type rteHelper interface {
 }
 
 type daemonSetManifest struct {
-	daemonSet      *appsv1.DaemonSet
-	daemonSetError error
+	daemonSet             *appsv1.DaemonSet
+	daemonSetError        error
+	rteComputedConfigHash string
 }
 
 type machineConfigManifest struct {
@@ -227,6 +228,7 @@ type GeneratedDesiredManifest struct {
 	NodeGroup         *nropv1.NodeGroup
 	// generated manifests
 	DaemonSet             *appsv1.DaemonSet
+	RTEComputedConfigHash string
 	IsCustomPolicyEnabled bool
 }
 ```

Tal-or committed:
cleanup: delete set annotation from render code

The annotation for hash of the configmap should be much later in the code.
We cannot and should not set the annotation outside the reconciliation loop,
since we don't have the needed inforamtion at this point in time.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani commented:
/retest

ffromani reviewed (commented):
/approve
/lgtm

thanks!
